src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+                    int stopIndex = (int)(ticks & this.owner.mask);
+
+                    HashedWheelBucket bucket = this.owner.wheel[stopIndex];
+                    bucket.AddTimeout(timeout);
+                }
+            }
+
+            void ProcessCancelledTasks()
+            {
+                while (true)
+                {
+                    HashedWheelTimeout timeout;
+                    if (!this.owner.cancelledTimeouts.TryDequeue(out timeout))
+                    {
+                        // all processed
+                        break;
+                    }
+                    try
+                    {
+                        timeout.Remove();
+                    }
+                    catch (Exception ex)
+                    {
+                        if (Logger.WarnEnabled)
+                        {
+                            Logger.Warn("An exception was thrown while processing a cancellation task", ex);
+                        }
+                    }
+                }
+            }
+
+            /// <summary>
+            /// calculate timer firing time from startTime and current tick number,
+            /// then wait until that goal has been reached.
+            /// </summary>
+            /// <returns>long.MinValue if received a shutdown request,
+            /// current time otherwise (with long.MinValue changed by +1)
+            /// </returns>
+            TimeSpan WaitForNextTick()
+            {
+                long deadline = this.owner.tickDuration * (this.tick + 1);
+
+                while (true)
+                {
+                    TimeSpan currentTime = (PreciseTimeSpan.FromStart - this.owner.StartTime).ToTimeSpan();
+                    TimeSpan sleepTime = CeilTimeSpanToMilliseconds(TimeSpan.FromTicks(deadline - currentTime.Ticks));
+
+                    if (sleepTime <= TimeSpan.Zero)
+                    {
+                        if (currentTime.Ticks == long.MinValue)
+                        {
+                            return TimeSpan.FromTicks(-long.MaxValue);
+                        }
+                        else
+                        {
+                            return currentTime;
+                        }
+                    }
+
+                    Task delay = null;
+                    try
+                    {
+                        long sleepTimeMs = sleepTime.Ticks / TimeSpan.TicksPerMillisecond; // we've already rounded so no worries about the remainder > 0 here
+                        Contract.Assert(sleepTimeMs <= int.MaxValue);
+                        delay = Task.Delay((int)sleepTimeMs, this.owner.CancellationToken);
+                        delay.Wait();
+                    }
+                    catch (AggregateException) when (delay != null && delay.IsCanceled)
+                    {
+                        if (Volatile.Read(ref this.owner.workerStateVolatile) == WorkerStateShutdown)
```

Comment:
IMO hypothetical bug chasing is not productive. and yes, it would.

---
In reply to: [114476824](https://github.com/Azure/DotNetty/pull/242#discussion_r114476824) [](ancestors = 114476824)